### ISSUE TITLE
fx:include creates cyclic references

### ISSUE THREAD

**Issue by jperedadnr** [gluonhq/substrate#291]
Using `<fx:include source="custom.fxml">` in a given `hello.fxml` file causes:

```
Caused by: javafx.fxml.LoadException: 
hellofx/custom.fxml:30
hellofx/hello.fxml:30

        at javafx.fxml.FXMLLoader.constructLoadException(FXMLLoader.java:2625)
        at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2603)
        at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2466)
        at javafx.fxml.FXMLLoader$IncludeElement.constructValue(FXMLLoader.java:1154)
        at javafx.fxml.FXMLLoader$ValueElement.processStartElement(FXMLLoader.java:754)
        at javafx.fxml.FXMLLoader.processStartElement(FXMLLoader.java:2722)
  ...
Caused by: java.io.IOException: Including "resource:hellofx/custom.fxml" in "resource:hellofx/custom.fxml" created cyclic reference.
        at javafx.fxml.FXMLLoader$IncludeElement.constructValue(FXMLLoader.java:1146)
        at javafx.fxml.FXMLLoader$ValueElement.processStartElement(FXMLLoader.java:754)
        at javafx.fxml.FXMLLoader.processStartElement(FXMLLoader.java:2722)
        at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2552)
``` 

Note that line 30 is the line in hello.fxml with the `<fx:include>` tag. 
However the printout `hellofx/custom.fxml:30` doesn't make sense (as in this case custom.fxml has a just few lines).

**Comment by jperedadnr** [comment]
How to reproduce:

BoxController.java
```
package hellofx;

import javafx.fxml.FXML;
import javafx.scene.control.Label;

public class BoxController {

    @FXML private Label label;

    public void initialize() {
        label.setText(System.getProperty("javafx.version"));
    }

}
```

HelloController.java
```
package hellofx;

import javafx.fxml.FXML;

import java.util.ResourceBundle;

public class HelloController {

    @FXML
    private BoxController customController;

    @FXML
    private ResourceBundle resources;

    public void initialize() {
    }

}
```

HelloFXML.java
```
package hellofx;

import javafx.application.Application;
import javafx.fxml.FXMLLoader;
import javafx.scene.Scene;
import javafx.scene.layout.AnchorPane;
import javafx.stage.Stage;

import java.io.IOException;

public class HelloFXML extends Application {

    @Override
    public void start(Stage stage) throws IOException {
        AnchorPane root = FXMLLoader.load(HelloFXML.class.getResource("hello.fxml"));
        Scene scene = new Scene(root, 640, 480);
        stage.setScene(scene);
        stage.show();
    }

    public static void main(String[] args) {
        launch(args);
    }

}
```
custom.fxml
```
<?xml version="1.0" encoding="UTF-8"?>

<?import javafx.scene.control.Button?>
<?import javafx.scene.control.Label?>
<?import javafx.scene.layout.VBox?>

<VBox fx:id="box" prefHeight="400.0" prefWidth="600.0" xmlns="http://javafx.com/javafx/11.0.1" xmlns:fx="http://javafx.com/fxml/1" fx:controller="hellofx.BoxController">
   <children>
            <Button fx:id="button" mnemonicParsing="false" text="click me" />
            <Label fx:id="label" text="a label" visible="false" />
   </children>
</VBox>
```

hello.fxml
```
<?xml version="1.0" encoding="UTF-8"?>
<?import javafx.scene.layout.AnchorPane?>

<AnchorPane fx:id="pane" prefHeight="400.0" prefWidth="600.0"  xmlns="http://javafx.com/javafx/11.0.1" xmlns:fx="http://javafx.com/fxml/1" fx:controller="hellofx.HelloController">
   <children>
      <fx:include source="custom.fxml" fx:id="custom"/>
   </children>
</AnchorPane>
```

pom.xml
```
<plugin>
                <groupId>com.gluonhq</groupId>
                <artifactId>client-maven-plugin</artifactId>
                <version>0.1.14-SNAPSHOT</version>
                <configuration>
                    <reflectionList>
                        <list>javafx.fxml.FXMLLoader</list>
                        <list>hellofx.HelloController</list>
                        <list>hellofx.BoxController</list>
                    </reflectionList>
                    <mainClass>hellofx.HelloFXML</mainClass>
                </configuration>
            </plugin>
```

**Comment by jperedadnr** [comment]
Typically we use one FXMLLoader instance per FXML file and that works fine, but in this case, we use the same FXMLLoader for the main FXML file and those referenced by it via fx:include.
I've narrowed down the issue to this line:
https://github.com/openjdk/jfx/blob/master/modules/javafx.fxml/src/main/java/javafx/fxml/FXMLLoader.java#L2503

Basically:
```
private XMLStreamReader xmlStreamReader = null;
...    
private <T> T loadImpl(InputStream inputStream, Class<?> callerClass) throws IOException {
    InputStreamReader inputStreamReader = new InputStreamReader(inputStream, charset);
    xmlStreamReader = new StreamReaderDelegate(xmlInputFactory.createXMLStreamReader(inputStreamReader)) {...}
    while (xmlStreamReader.hasNext()) {
                ...
    }
}            
```

With Substrate the inputStream is correctly set to each FXML file, however it seems xmlStreamReader is somehow cached, and xmlStreamReader.hasNext() processes always the initial content (the main file).

We should try to isolate the case, without JavaFX, with a nested xmlStreamReader in HelloWorld

**Comment by tadschik** [comment]
Any workaround possible?

**Comment by sgyeme** [comment]
I am stack on this same issue. Any work arround

**Comment by sgyeme** [comment]
Another work arround that has word for me is substitution as found bellow

``import java.io.IOException;
import java.io.InputStream;
import java.net.URL;

import com.oracle.svm.core.annotate.Alias;
import com.oracle.svm.core.annotate.Substitute;
import com.oracle.svm.core.annotate.TargetClass;

import javafx.fxml.FXMLLoader;

@TargetClass(FXMLLoader.class)
public final class FXML2Loader {
	
	@Alias private URL location;
	@Substitute
	private <T> T loadImpl(final Class<?> callerClass) throws IOException {
		if (location == null) {
			throw new IllegalStateException("Location is not set.");
		}
		InputStream inputStream = null;
		T value;
		try {
			inputStream = new URL(location.toExternalForm()).openStream();
			value = loadImpl(inputStream, callerClass);
		} finally {
			if (inputStream != null) {
				inputStream.close();
			}
		}

		return value;
	}
   @Alias
	@SuppressWarnings("unchecked")
	private <T> T loadImpl(InputStream inputStream, Class<?> callerClass) {
		var root = new Object();
		return (T) root;

	}
}

**Comment by jperedadnr** [comment]
Should be fixed with Client plugin 0.1.36-SNAPSHOT

Requires snapshots repo (until new release):
```
 <pluginRepositories>
        <pluginRepository>
            <id>snapshots</id>
            <url>http://oss.sonatype.org/content/repositories/snapshots</url>
        </pluginRepository>
    </pluginRepositories>
```
